src/main/java/org/opentripplanner/standalone/config/framework/json/ParameterBuilder.java

```diff
@@ -0,0 +1,571 @@
+  }
+
+  private NodeAdapter buildObject() {
+    JsonNode node = target.addAndValidateParameterNode(info.build());
+    return target.path(paramName(), node);
+  }
+
+  private <T> List<T> buildAndListValues(List<T> defaultValues, Function<JsonNode, T> mapper) {
+    JsonNode array = build();
+    if (array.isMissingNode()) {
+      return defaultValues;
+    }
+    if (!array.isArray()) {
+      throw error("The parameter is not a JSON array as expected.");
+    }
+    List<T> values = new ArrayList<>();
+    for (JsonNode node : array) {
+      values.add(mapper.apply(node));
+    }
+    return values;
+  }
+
+  private <T> List<T> buildAndListObjects(List<T> defaultValues, Function<NodeAdapter, T> parse) {
+    var array = build();
+    if (array.isMissingNode()) {
+      return defaultValues;
+    }
+    if (!array.isArray()) {
+      throw error("The parameter is not a JSON array as expected.");
+    }
+    List<T> values = new ArrayList<>();
+    int i = 0;
+    for (JsonNode node : array) {
+      values.add(parse.apply(target.path("[" + i + "]", node)));
+      ++i;
+    }
+    return values;
+  }
+
+  /* parse/map from string to a specific type, handle parse error */
+
+  private <T> T parseConfigType(ConfigType elementType, JsonNode child) {
+    try {
+      return ConfigType.getParameter(elementType, child);
+    } catch (Exception e) {
+      throw error(
+        "The parameter value '%s' is not of type %s.".formatted(
+            child.asText(),
+            elementType.docName()
+          ),
+        e
+      );
+    }
+  }
+
+  private <T extends Enum<T>> T parseEnum(String value, Class<T> ofType) {
+    var upperCaseValue = value.toUpperCase();
+    return Stream
+      .of(ofType.getEnumConstants())
+      .filter(it -> it.name().toUpperCase().equals(upperCaseValue))
+      .findFirst()
+      .orElseThrow(() -> {
+        throw error(
+          "The parameter value '%s' is not in legal. Expected one of %s.".formatted(
+              value,
+              List.of(ofType.getEnumConstants())
+            )
+        );
+      });
+  }
+
+  private LocalDate parseRelativeLocalDate(String value, ZoneId zoneId) {
+    try {
+      return LocalDateUtils.asRelativeLocalDate(value, LocalDate.now(zoneId));
+    } catch (DateTimeParseException e) {
+      throw error("The parameter value '%s' is not a Period or LocalDate.".formatted(value), e);
+    }
+  }
+
+  private Duration parseDuration(String value) {
+    try {
+      return DurationUtils.duration(value);
+    } catch (DateTimeParseException e) {
+      throw error("The parameter value '%s' is not a duration.".formatted(value), e);
+    }
+  }
+
+  private Duration parseDuration(String value, ChronoUnit unit) {
+    try {
+      return DurationUtils.duration(value, unit);
+    } catch (DateTimeParseException e) {
+      throw error("The parameter value '%s' is not a duration.".formatted(value), e);
+    }
+  }
+
+  private Locale parseLocale(String text) {
```

Comment:
Should we add some comment(s) here as the code is a bit cryptic?

Reply:
I added JavaDoc, I hope that clearify it.